backend/data/users.js

```diff
@@ -252,76 +248,346 @@ async function getCourses(userId) {
+  if (!chunksQuery || chunksQuery.deletedCount === 0) throw new Error(`Failed to delete assignment chunks`);
+
+  return true;
+}
+
+/**
+ * Deletes all instances of file data given a fileId
+ * (chunks and metadata included)
+ * @param {ObjectId} fileId
+ * @returns boolean
+ */
+async function deleteFileInstanceById(fileId) {
+  const parsedFileId = error.validId(fileId);
+
+  // First we need to delete the file metadata
+  const assignmentsFilesCollection = await assignmentsFiles();
+  const filesQuery = await assignmentsFilesCollection.deleteMany({ _id: parsedFileId });
+  if (!filesQuery || filesQuery.deletedCount === 0) return false;
+
+  // Then delete the file metadata
+  const assignmentsChunksCollection = await assignmentsChunks();
+  const chunksQuery = await assignmentsChunksCollection.deleteMany({ files_id: parsedFileId });
+  if (!chunksQuery || chunksQuery.deletedCount === 0) return false;
+
+  return true;
+}
+
+/**
+ * Updates database with new submission file ID
+ * Also overwrites existing submission fileIDs when applicable
+ * @param {ObjectId} studentId
+ * @param {ObjectId} assignmentId
+ * @param {ObjectId} fileId
+ * @returns { overwritten: boolean, modifiedCount: number }
+ */
+async function submitAssignment(studentId, assignmentId, fileId) {
+  const parsedStudentId = error.validId(studentId);
+  const parsedAssignmentId = error.validId(assignmentId);
+  const parsedFileId = error.validId(fileId);
+
+  // 1: Ensure the user is a student
+  const userObj = await getUser(parsedStudentId);
+  if (!userObj || userObj?.role !== 'student') throw new Error('Students are only allowed to upload assignments');
+
+  // 2: Ensure assignment is in a course
+  const coursesCollection = await courses();
+  const courseQuery = await coursesCollection.findOne({ 'assignments._id': parsedAssignmentId });
+  if (!courseQuery) throw new Error('No course found for provided assignment id');
+
+  let modified = false;
+  let overwritten = false;
+  let matchedIndex = 0;
+
+  for (let i = 0; i < userObj.classes.length; i++) {
+    let currClass = userObj.classes[i];
+    if (currClass._id.toString() === courseQuery._id.toString()) {
+      // Check that the student's enrolled course matches the assignment's course
+      matchedIndex = i;
+      for (let j = 0; j < currClass.grades.length; j++) {
+        // Check if assignment already has a submission
+        let currGrade = currClass.grades[j];
+        if (currGrade._id.toString() === parsedAssignmentId.toString()) {
+          // Delete all necessary old file data from assignments collection
+          overwritten = await submissionExists(parsedStudentId, parsedAssignmentId);
+          currGrade.submissionFile = parsedFileId;
+        }
+      }
+      modified = true;
+    }
+  }
+
+  // 3: Ensure student is enrolled in course
+  if (!modified) throw new Error('Unable to add new submission due to no student enrolled');
```

Comment:
Minor thing but still threw me off during testing, might want to rephrase this as "...due to student not enrolled in this assignment's course".

When I attempted to upload a submission to an assignment of a course that the current student account was not enrolled in, got this response and was thrown off by phrasing.